shlr/sdb/src/ht.c

```diff
@@ -1,239 +1,258 @@
+// calcsize - function to calculate the size of a value. if NULL, just stores 0.
+static SdbHash* internal_ht_new(ut32 size, HashFunction hashfunction, ListComparator comparator, DupKey keydup, DupValue valdup, HtKvFreeFunc pair_free, CalcSize calcsize) {
+	SdbHash* ht = calloc (1, sizeof (*ht));
+	if (!ht) {
+		return NULL;
 	}
-	return NULL;
+	ht->size = size;
+	ht->count = 0;
+	ht->prime_idx = 0;
+	ht->load_factor = 1;
+	ht->hashfn = hashfunction;
+	ht->cmp = comparator;
+	ht->dupkey = keydup;
+	ht->dupvalue = valdup;
+	ht->table = calloc (ht->size, sizeof (SdbList*));
+	ht->calcsize = calcsize;
+	ht->freefn = pair_free;
+	ht->deleted = ls_newf (free);
+	// Because we use calloc, each listptr will be NULL until used */
+	return ht;
 }
 
-static int rehash = 0;
-static void ht_rehash(SdbHash *ht, ut32 new_size_index) {
-	SdbHash old_ht = *ht;
-	SdbHashEntry *e;
-	if (!ht || new_size_index >= ARRAY_SIZE (hash_sizes))
-		return;
-	// XXX: This code is redupped! fuck't
-	ht->table = calloc (hash_sizes[new_size_index].size, sizeof (*ht->table));
-	if (!ht->table)
-		return;
-	rehash = 1;
-	ht->size_index = new_size_index;
-	ht->size = hash_sizes[ht->size_index].size;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	for (e = old_ht.table; e != old_ht.table + old_ht.size; e++) {
-		if (entry_is_present (e))
-			ht_insert (ht, e->hash, e->data, e->iter);
+bool ht_delete_internal(SdbHash* ht, const char* key, ut32* hash) {
+	SdbKv* kvp;
+	SdbListIter* iter;
+	ut32 computed_hash = hash ? *hash : ht->hashfn (key);
+	ut32 bucket = computed_hash % ht->size;
+	SdbList* list = ht->table[bucket];
+	ls_foreach (list, iter, kvp) {
+		if (key == kvp->key || (ht->cmp && !ht->cmp (key, kvp->key))) {
+#define EXCHANGE 1
+#if EXCHANGE
+			ls_split_iter (list, iter);
+			ls_append (ht->deleted, iter);
+			iter->data = NULL;
+#else
+			ls_delete (list, iter);
+#endif
+			ht->count--;
+			return true;
+		}
 	}
-	free (old_ht.table);
-rehash = 0;
+	return false;
 }
 
-SdbHash* ht_new(SdbListFree f) {
-	SdbHash *ht = R_NEW (SdbHash);
-	if (!ht) return NULL;
-	// TODO: use slices here
-	ht->list = ls_new ();
-	ht->list->free = f;
-	ht->size = hash_sizes[0].size;
-	ht->table = calloc (ht->size, sizeof (*ht->table));
-	if (!ht->table) {
-		free (ht);
-		return NULL;
-	}
-	ht->size_index = 0;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	return ht;
+SdbHash* ht_new() {
+	return internal_ht_new (ht_primes_sizes[0], sdb_hash, strcmp, strdup,
+			strdup, sdb_kv_free, strlen);
 }
 
-void ht_free(SdbHash *ht) {
-	if (ht) {
-		free (ht->table);
-		ls_free (ht->list);
-		free (ht);
+void ht_free(SdbHash* ht) {
+	ut32 i;
+	for (i = 0; i < ht->size; i++) {
+		ls_free (ht->table[i]);
 	}
+	ht_free_deleted (ht);
```

Comment:
i think this leaks memory, because `ht->deleted` is re-allocated in `ht_free_deleted`.